### PR TITLE
[WIP] feat: support k8s multiple apiservers

### PR DESCRIPTION
**Pull request by bittermandel**
PSA: This is a Draft PR to keep track of progress and get early comments.
 
This PR is a continuation of https://github.com/cilium/cilium/pull/20090/.

With the upstream changes since that PR was worked on, this PR had to do things slightly different.

Short explanation:

The config used for the Kubernetes client is wrapped in a Roundtripper that rotates the Host based on a list passed through the config field `k8s-api-server-urls`. While establishing the initial connection, it rotates through the APIServerURLs one by one until it manages to establish connection.

Features missing:
* Always rotate URL if the client fails to connect for X time, not only during startUp.
* When initial connection has been established, replace the apiserver urls passed through config with the endpoints of the `kubernetes` service.

### CODE REVIEW

File: pkg/k8s/client/cell.go [hunk 1]
```diff
@@ -109,29 +112,69 @@ type compositeClientset struct {
 	log           logrus.FieldLogger
 	closeAllConns func()
 	restConfig    *rest.Config
+	rt            *rotatingHttpRoundTripper
+
+	lock.RWMutex
```
Comment (bittermandel) [inline pkg/k8s/client/cell.go:117]: I was not able to put this lock in `cell.Config` due to Cell assuming that all struct fields are passed by pflags. Not sure if there is a better place for it than here.

File: pkg/k8s/client/cell.go [hunk 2]
```diff
@@ -109,29 +112,69 @@ type compositeClientset struct {
...
+	} else {
+		apiServerURL, err := url.Parse(restConfig.Host)
+		if err != nil {
+			return nil, err
+		}
+		client.rt.apiServerURL = apiServerURL
+		log.Infof("Setting APIServerURL to %s", apiServerURL)
+	}
+
+	restConfig.Wrap(client.WrapRoundTripper)
```
Comment (bittermandel) [inline pkg/k8s/client/cell.go:181]: This makes rotatingHttpRoundTripper.Roundtrip run before the pre-configured one. This enables us to change the receiving Host dynamically.

**Comment by squeed** [comment]
Sweet.

What is the best-practice for kube-proxy and kubelet here? AFAICT, kubelet still doesn't support multiple api servers; is there a plan for that to change as well?

**Comment by brb** [comment]
> What is the best-practice for kube-proxy and kubelet here

* kube-proxy - use Cilium's kube-proxy replacement :smile: 
* kubelet - how does apiserver URL gets specified to it? I guess it doesn't use ClusterIP of the `kubernetes` service due to the :chicken: :egg:  problem?

**Comment by brb** [comment]
@bittermandel  bump ^^

**Comment by bittermandel** [comment]
Sorry, I took a bit of a detour!

After some considerations, I am not sure this feature makes sense since the Kubelet has to be configured with a single Kubeconfig cluster as well. Even kubeadm join gets its apiserver endpoint from the `kube-public/cluster-info` config map, which is also a single Kubeconfig cluster.

To achieve resilience and reliable connectivity, we ended up relying on kube-vip for the kubelet as well as Cilium. Until the Kubelet can also support multiple apiservers, I don't see how this feature can be deployed and relied upon fully. See the very long and old discussion here: https://github.com/kubernetes/kubernetes/issues/18174.

TLDR: When using kubeProxyReplacement, we will have to rely on the same address that is used by the Kubelet. The Kubelet  does not support multiple apiservers, so that will have to be solved in tangent.
